packages/wallet/src/redux/protocols/ledger-top-up/reducer.ts

```diff
@@ -0,0 +1,288 @@
+    }
+    return { protocolState: states.waitForPostTopUpLedgerUpdate(protocolState), sharedData };
+  }
+
+  return { protocolState, sharedData };
+};
+
+const waitForPostTopUpLedgerUpdateReducer: ProtocolReducer<states.LedgerTopUpState> = (
+  protocolState: states.WaitForPostTopUpLedgerUpdate,
+  sharedData: SharedData,
+  action: LedgerTopUpAction,
+): ProtocolStateWithSharedData<states.LedgerTopUpState> => {
+  if (action.type !== 'WALLET.COMMON.COMMITMENT_RECEIVED') {
+    console.warn(
+      `Ledger Top Up Protocol expected COMMITMENT_RECEIVED received ${action.type} instead.`,
+    );
+    return { protocolState, sharedData };
+  }
+
+  const checkResult = checkAndStore(sharedData, action.signedCommitment);
+
+  if (!checkResult.isSuccess) {
+    return {
+      protocolState: states.failure({ reason: 'Received Invalid Commitment' }),
+      sharedData,
+    };
+  }
+  sharedData = checkResult.store;
+  const isFirstPlayer = helpers.isFirstPlayer(protocolState.ledgerId, sharedData);
+  // Accept consensus if player B
+  if (!isFirstPlayer) {
+    const ourCommitment = acceptConsensus(action.signedCommitment.commitment);
+    const signResult = signAndStore(sharedData, ourCommitment);
+    if (!signResult.isSuccess) {
+      return { protocolState: states.failure({ reason: 'Signature Failure' }), sharedData };
+    }
+    sharedData = signResult.store;
+
+    const messageRelay = sendCommitmentReceived(
+      getOpponentAddress(sharedData, protocolState.ledgerId),
+      protocolState.processId,
+      signResult.signedCommitment.commitment,
+      signResult.signedCommitment.signature,
+      LEDGER_TOP_UP_PROTOCOL_LOCATOR,
+    );
+    sharedData = queueMessage(sharedData, messageRelay);
+  }
+
+  return { protocolState: states.success({}), sharedData };
+};
+
+function calculateTotalTopUp(allocation: string[]): string {
+  return bigNumberify(allocation[2])
+    .add(allocation[3])
+    .toHexString();
+}
+function getOpponentAddress(sharedData: SharedData, channelId: string) {
+  const channel = getChannel(sharedData, channelId);
+  if (!channel) {
+    throw new Error(`Could not find channel for id ${channelId}`);
+  }
+
+  return theirAddress(channel);
+}
+function getLatestCommitment(sharedData: SharedData, channelId: string) {
+  const channel = getChannel(sharedData, channelId);
+  if (!channel) {
+    throw new Error(`Could not find channel for id ${channelId}`);
+  }
+
+  return getLastCommitment(channel);
+}
+function calculateTopUpAllocation(
+  proposedAllocation: string[],
+  currentAllocation: string[],
+): string[] {
+  const newAllocation: string[] = [];
+  for (let i = 0; i < currentAllocation.length; i++) {
+    const currentFunds = bigNumberify(currentAllocation[i]);
+    const proposedFunds = bigNumberify(proposedAllocation[i]);
+    const topUp = currentFunds.gte(proposedFunds)
+      ? '0x0'
+      : proposedFunds.sub(currentFunds).toHexString();
+    newAllocation[i] = topUp;
+  }
+  return newAllocation;
+}
```

Comment:
I feel like the accounting logic is complicated enough to merit unit testing.

Reply:
Deferring until #543